src/adapter.jl

```diff
@@ -264,3 +264,16 @@ end
 # `false` if the default `hash` is used.
 Base.hash(::GapObj, h::UInt) = h
 Base.hash(::FFE, h::UInt) = h
+
+# iteration
+
+function Base.iterate(obj::GapObj)
+    Globals.IsList(obj) || Globals.IsCollection(obj) ||
+        throw(ArgumentError("object cannot be iterated"))
+    iterate(obj, (1, Globals.Length(obj)))
+end
+
+function Base.iterate(obj::GapObj, (i, len))
+    i > len && return nothing
+    ElmList(obj, i), (i+1, len)
```

Comment:
ElmList is only suitable for lists, not for non-list collections. For those you really need an iterator.

For example, take `GAP.Globals.SymmetricGroup(3)`

Reply:
Yes I'm awayre it's not complete. For now I just copied the method in the `Set` constructor from Thomas PR, which uses `ElmList`, and accepts only `Globals.IsList(obj) || Globals.IsCollection(obj) `. I didn't add for example ``Globals.IsVectorObj` as I didn't know yet how to construct and test such object.

Reply:
so the way to do general iteration is GAP is to use ... an iterator ;-). But that comes with an overhead, esp. since GAP cannot optimize the iterator object away, so it's desirable to use `ElmList` when possible (for `IsList`), but otherwise use the iterator.

In GAP one can write `for x in coll do ... od` which is syntactic sugar for roughly this:
```gap
  iter := Iterator(coll)
  while not IsDoneIterator(iter) do
    x := NextIterator(iter);
  od;
```

Reply:
So I rebased, but still without including general iteration yet, I will look into that now.

Reply:
Seems you've done it :-)